src/sig_variant.zig

```diff
@@ -643,60 +649,152 @@ pub fn createSigVariant(
             }
 
-            // TODO - check msg uniqueness?
-
-            if (pairing_buffer.len < Pairing.sizeOf()) {
-                return BLST_ERROR.FAILED_PAIRING;
-            }
-
-            var pairing = Pairing.new(&pairing_buffer[0], pairing_buffer.len, hash_or_encode, &dst[0], dst.len) catch return BLST_ERROR.FAILED_PAIRING;
-
-            for (0..n_elems) |i| {
-                try pairing.mulAndAggregate(&pks[i].point, pks_validate, &sigs[i].point, sigs_groupcheck, &rands[i][0], rand_bits, &msgs[i][0], msgs[i].len, null);
+            const AtomicCounter = std.atomic.Value(usize);
+            // donot use AtomicBoolean because we want error code to return
+            const AtomicError = std.atomic.Value(c_uint);
+            var atomic_counter = AtomicCounter.init(0);
+            // 0 = BLST_SUCCESS
+            var atomic_valid = AtomicError.init(0);
+            var wg = std.Thread.WaitGroup{};
+
+            const cpu_count = @max(1, std.Thread.getCpuCount() catch 1);
+            const n_workers = @min(cpu_count, n_elems);
+            const Signature = @This();
+
+            for (0..n_workers) |_| {
+                spawnTaskWg(&wg, struct {
+                    fn run(_msgs: [][]const u8, _dst: []const u8, _pks: []const *PublicKey, _pks_validate: bool, _sigs: []const *Signature, _sigs_groupcheck: bool, _rands: [][]const u8, _rand_bits: usize, _pool: *MemoryPool, _atomic_counter: *AtomicCounter, _atomic_valid: *AtomicError) void {
+                        const pairing_buffer = _pool.getPairingBuffer() catch {
+                            // .release will publish the value to other threads
+                            _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            return;
+                        };
+                        defer {
+                            _pool.returnPairingBuffer(pairing_buffer) catch {
+                                // .release will publish the value to other threads
+                                _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            };
+                        }
+                        var pairing = Pairing.new(&pairing_buffer[0], pairing_buffer.len, hash_or_encode, &_dst[0], _dst.len) catch {
+                            // .release will publish the value to other threads
+                            _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            return;
+                        };
+
+                        // the most relaxed atomic ordering
+                        var local_count: usize = 0;
+                        while (_atomic_valid.load(.monotonic) == 0) {
+                            // this uses @atomicRmw internally and returns the previous value
+                            // acquired then release which publish value to other thread
+                            const counter = _atomic_counter.fetchAdd(1, AtomicOrder.acq_rel);
```

Comment:
Rust binding always use "Relaxed" atomic order which could be a bit faster. It's worth to review and try it later.